src/cppnv/EnvReader.cc

```diff
@@ -0,0 +1,813 @@
+#include "EnvReader.h"
+
+
+namespace cppnv {
+EnvReader::read_result EnvReader::read_pair(EnvStream* file,
+                                            const EnvPair* pair) {
+  const read_result result = read_key(file, pair->key);
+  if (result == fail || result == empty) {
+    return fail;
+  }
+  if (result == comment_encountered) {
+    return comment_encountered;
+  }
+  if (result == end_of_stream_key) {
+    return end_of_stream_key;
+  }
+
+  if (result == end_of_stream_value) {
+    return success;
+  }
+  //  trim right side of key
+  while (pair->key->key_index > 0) {
+    if (pair->key->key->at(pair->key->key_index - 1) != ' ') {
+      break;
+    }
+    pair->key->key_index--;
+  }
+  if (!pair->key->has_own_buffer()) {
+    const auto tmp_str = new std::string(pair->key->key_index, '\0');
+
+    tmp_str->replace(0,
+                     pair->key->key_index,
+                     *pair->key->key,
+                     0,
+                     pair->key->key_index);
+
+    pair->key->set_own_buffer(tmp_str);;
+  } else {
+    pair->key->clip_own_buffer(pair->key->key_index);
+  }
+  pair->value->value->clear();
+  const read_result value_result = read_value(file, pair->value);
+  if (value_result == end_of_stream_value) {
+    return end_of_stream_value;
+  }
+  if (value_result == comment_encountered || value_result == success) {
+    if (!pair->value->has_own_buffer()) {
+      const auto tmp_str =
+          new std::string(pair->value->value_index, '\0');
+
+      tmp_str->replace(0,
+                       pair->value->value_index,
+                       *pair->value->value,
+                       0,
+                       pair->value->value_index);
+
+      pair->value->
+            set_own_buffer(tmp_str);
+    } else {
+      pair->value->clip_own_buffer(pair->value->value_index);
+    }
+    remove_unclosed_interpolation(pair->value);
+    return success;
+  }
+  if (value_result == empty) {
+    remove_unclosed_interpolation(pair->value);
+    return empty;
+  }
+  if (value_result == end_of_stream_key) {
+    remove_unclosed_interpolation(pair->value);
+    return end_of_stream_key;
+  }
+
+  remove_unclosed_interpolation(pair->value);
+  return fail;
+}
+
+
+int EnvReader::read_pairs(EnvStream* file, std::vector<EnvPair*>* pairs) {
+  int count = 0;
+  auto buffer = std::string(256, '\0');
+
+  while (true) {
+    buffer.clear();
+    EnvPair* pair = new EnvPair();
```

Comment:
these should make use of `std::unique_ptr` instead of the explicit new/delete.

Reply:
I'm not sure if they would benefit from this. They are only deleted if the read fails and then held onto until they can be transferred into the env store map later. Since there additional benefit is the additional overhead required? Please make it clearer.

Reply:
> Since there additional benefit is the additional overhead required?

What overhead? Manual memory management seems undesirable here, and moving `unique_ptr` objects around has virtually no overhead.

Reply:
https://youtu.be/rHIkrotSwcc?t=1266

Reply:
I said "virtually no overhead", not "zero overhead".

Reply:
Can you explain what kind of exception will happen in a string parser that the overhead is worth?

Reply:
I believe that automatic memory management is widely accepted as a best practice in modern C++, and it should require no justification. Deviating from that and using manual memory management instead is what requires justification, and potentially saving a handful of CPU cycles when Node.js starts does not seem like a strong argument to me.

Reply:
We're not worried about ownership. We're not worried about these object's lifetime as they only exist briefly and we'd want to free them. We're not writing exception prone code so the 6% increase in speed typically seen between raw pointers and unique_ptr, more than a handful of cycles, during start up is eventually saving compute hours. 

I'm not against using unique_ptr or other ownership or automatic memory management idioms -- I was simply writing for speed. Using something that is slower goes against the grain of having reused string instead of willy-nilly allocating them as needed. The justification is this whole project was written, including writing my own stream facade to get a 10 fold speed increase, for speed. 

The bun parser is nearly as fast but doesn't correctly do control codes or here docs. Env files are based on bash so heredocs and interpolation, to answer another one of your questions, are the standard. We do them correctly. We do them faster than bun. Adding lag because what was once the gateway to entry, memory management, has been lessened is not in spirit with the reason the code was written.